all-sdk/autotuning.md

```diff
@@ -0,0 +1,431 @@
+  logic and contact support about it.
+* More complexity in the core/lang bridge, nothing too bad though.
+
+## Alternatives
+
+We could not expose this as a user-implementable interface at all and only offer a handful of default implementations.
+The benefits of user customization are substantial though, and seem to outweigh the mild complexity increase. Users will
+always know more about their workload than we can.
+
+Specifically concerning `mark_slot_used`, we could potentially allow an error to be returned here, causing the SDK to
+immediately fail the task instead of working on it. This is maybe a nice way to let users reject tasks they don't want
+this worker to handle... however there are some obvious potential thrashing issues there (ex: only one worker rejecting
+the same task over and over). Allowing this seems like it'd mostly be a way for users to bandaid over something that was
+really a mistake in the way their task queues are architected.
+
+## Adoption & Teaching
+
+We should roll out the feature by introducing the interface & reimplementing the existing behavior behind that
+interface. There will be no visible default behavior changes. Likely we'll include some other OOTB implementations like
+a memory-based one that users can opt-in to.
+
+We might, at some point, decide that the default implementation should change from the fixed-number-of-slots impl to
+a more dynamic one, if we find that it substantially outperforms in most scenarios. If and when we do make that change,
+we'll want to give users some advance warning, since although the change is not technically breaking, it could have 
+knock-on effects in terms of the external load on the systems their workers are interacting with.
+
+Docs & education will both need content about what the interface is for, what default implementations are provided &
+what their limitations are, and how to implement a custom one.
+
+## Open questions
+
+### Async/fallible `mark_slot_used` and `release_slot`
+The "fun example" raises questions about whether `mark_slot_used` and `release_slot` should be allowed to be async
+and/or return errors. Users might reasonably want to perform side effects in them, but internal usage of these functions
+is ideally fast and infallible.
+
+If we were to allow them to be async/blocking, it's another possible spot things can get hosed up and stuck. Less than
+great. If we allow them to be fallible, it's not clear what we could do on a returned error besides bubble it up and
+shut down the worker.
+
+The downsides seem substantial, and users who want to do this can "offload" the async-ness/fallibility to the
+`reserve_slot` function as implied in my toy, which seems reasonable.
+
+## Future considerations
+
+This proposal has focused specifically on worker autotuning and user customization thereof. However, it's worth
+considering how this effort will fit in to future work we know we want to do, namely around autoscaling and on-demand
+workers.
```

Comment:
Another future option worth considering (imo) would be where some workers (at least new ones) would learn from already running workers about their learned configuration to speed up the learning process.